File: cmd/hubctl/Problem.go
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"sort"
	"strings"

	port "github.com/Jersyfi/hubtask/core/port/httpclient"
	"github.com/Jersyfi/hubtask/presentation/openapi"
)

// requestIDParam is the parameter `errors.internal` renders. The server puts the request ID in a
// field of its own rather than in the parameters, so it is merged in here - a message asking the
// reader to quote a reference has to have the reference in it.
const requestIDParam = "request_id"

// APIError is a refusal from the installation, already turned into a sentence.
//
// The rendering happens once, when the problem arrives, rather than in Error(): what a client
// shows a person is built from the code and the parameters (ADR-0011), and doing that at the
// point where the catalogue and the response are both in hand keeps the two from drifting apart.
type APIError struct {
	Status     int
	Code       string
	DetailCode string
	RequestID  string

	message string
	fields  []string
}

func (e APIError) Error() string {
	if len(e.fields) == 0 {
		return e.message
	}
	return e.message + "\n  " + strings.Join(e.fields, "\n  ")
}

// problem turns a 4xx or 5xx into an APIError.
//
// An answer that is not a problem document is reported by its status and nothing else. Printing
// the body would be printing whatever a proxy, a captive portal or a load balancer decided to
// send - which is HTML at best and misleading at worst.
func (c *Client) problem(response port.Response) error {
	contentType := header(response.Header, "Content-Type")
	if !strings.HasPrefix(contentType, "application/problem+json") {
		return fmt.Errorf("the installation answered %s (%s), which is not a problem document",
			statusText(response.Status), describe(contentType))
	}

	var document openapi.Problem
	if err := json.Unmarshal(response.Body, &document); err != nil {
		return fmt.Errorf("the installation answered %s with a problem document that cannot be read: %w",
			statusText(response.Status), err)
	}

	failure := APIError{
		Status:     document.Status,
		Code:       document.Code,
		DetailCode: value(document.DetailCode),
		RequestID:  value(document.RequestId),
	}

	params := stringParams(document.Params)
	if failure.RequestID != "" {
		if _, taken := params[requestIDParam]; !taken {
			params[requestIDParam] = failure.RequestID
		}
	}

	// The detail code is the specific one - `containers.parent_not_found` rather than
	// `not_found` - so it wins where the catalogue knows it. The contract code is the fallback,
	// and a code the catalogue has never heard of is printed as itself: a client that showed
	// nothing would leave the reader with a blank line and an exit code.
	message, known := c.message(failure.DetailCode, params)
	if !known {
		message, known = c.message("errors."+failure.Code, params)
	}
	if !known {
		message = firstNonEmpty(failure.DetailCode, failure.Code, statusText(document.Status))
	}
	failure.message = message

	if document.FieldErrors != nil {
		for _, field := range *document.FieldErrors {
			code := value(field.Code)
			text, ok := c.message(code, stringParams(field.Params))
			if !ok {
				text = code
			}
			failure.fields = append(failure.fields, strings.TrimPrefix(value(field.Path), "/")+": "+text)
		}
		sort.Strings(failure.fields)
	}
	return failure
}

// stringParams flattens the parameters of a problem document. They are strings on the wire - the
// domain's Params are map[string]string - and typed as `any` only because the generated schema
// says "additionalProperties". Anything else is formatted rather than dropped.
func stringParams(params *map[string]any) map[string]string {
	flattened := map[string]string{}
	if params == nil {
		return flattened
	}
	for name, raw := range *params {
		if text, ok := raw.(string); ok {
			flattened[name] = text
			continue
		}
		flattened[name] = fmt.Sprint(raw)
	}
	return flattened
}

func header(headers map[string][]string, name string) string {
	for key, values := range headers {
		if strings.EqualFold(key, name) && len(values) > 0 {
			return values[0]
		}
	}
	return ""
}

func statusText(status int) string {
	if text := http.StatusText(status); text != "" {
		return fmt.Sprintf("%d %s", status, text)
	}
	return fmt.Sprintf("%d", status)
}

func describe(contentType string) string {
	if contentType == "" {
		return "no content type"
	}
	return contentType
}

func value(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}

func firstNonEmpty(candidates ...string) string {
	for _, candidate := range candidates {
		if candidate != "" {
			return candidate
		}
	}
	return ""
}

File: apps/webapp/src/lib/problem.ts
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

/**
 * A problem document, turned into something a person can read and act on.
 *
 * RFC 9457 with this API's four additions (ADR-0025, `api-guidelines.md` §3): `code` is the
 * category, `detail_code` is what actually happened, `params` fills the sentence, `field_errors[]`
 * says which input is wrong, and `request_id` is what a support request is answered by. This
 * module is where those become one sentence, a message per field, and a reference to quote.
 *
 * Three decisions live here rather than in a component.
 *
 * **The specific code wins, if there is a message for it.** `validation_failed` is true and
 * useless; `items.title_too_long` is the same failure said usefully. But a `detail_code` the
 * catalogue has never heard of is worse than the category, so the choice is made by asking the
 * catalogue rather than by preferring blindly.
 *
 * **A field error belongs under its field.** The server sends a `path`; a form that showed the
 * five sentences in one list at the top would make the reader match them to twelve inputs by hand.
 *
 * **An internal error carries its reference.** Without it a support thread cannot be traced, and
 * the reader is the only person who has the number.
 */

import type { TransportError } from '@hubtask/sync-engine';

import type { MessageParams } from './i18n/format.ts';
import type { Messages } from './i18n/messages.ts';

export interface RenderedProblem {
  /** The one sentence to show. Never empty: every branch below ends in a message. */
  readonly message: string;
  /** Path from `field_errors[]` to the sentence for it, for a form to put beside its inputs. */
  readonly fields: ReadonlyMap<string, string>;
  /**
   * The request id, when there is one that is not already in the sentence. Shown where a person
   * can copy it - an internal error without its reference is a support thread nobody can trace.
   */
  readonly reference?: string;
  /** Whether this is the server's fault. What decides whether a retry is offered as the answer. */
  readonly isServerFault: boolean;
}

/**
 * What a failure that never reached the server says. These are conditions of the client's own
 * making, so they have no server code - and they are codes in the one catalogue rather than
 * sentences written here (ADR-0011).
 */
const KIND_CODES = {
  timeout: 'app.request_timed_out',
  offline: 'app.offline',
  malformed: 'app.answer_unreadable',
} as const;

/** The codes to try, most specific first. */
function candidates(error: TransportError): string[] {
  const out: string[] = [];
  if (error.detailCode) out.push(error.detailCode);
  if (error.code) {
    // The contract's codes are bare - `validation_failed` - and the catalogue keys them under
    // `errors.`; a code that already carries its group is left alone.
    out.push(error.code.includes('.') ? error.code : `errors.${error.code}`);
  }
  if (error.kind !== 'problem') out.push(KIND_CODES[error.kind]);
  // The floor, and which floor depends on whether there is a reference to quote. `errors.internal`
  // is the server's sentence and it *contains* the request id - offering it without one puts a
  // literal `{request_id}` on the screen, which is what a gateway answering 502 with an empty body
  // does. That case gets a sentence of its own instead.
  out.push(error.requestId ? 'errors.internal' : 'app.something_went_wrong');
  return out;
}

export function renderProblem(error: TransportError, messages: Messages): RenderedProblem {
  // The request id travels as a parameter as well as a field, because `errors.internal` renders it
  // inside its own sentence. Where it does, showing it again beneath would say it twice.
  const params: MessageParams = { ...error.params, ...(error.requestId ? { request_id: error.requestId } : {}) };

  const code = candidates(error).find((candidate) => messages.has(candidate)) ?? 'errors.internal';
  const message = messages.t(code, params);

  const fields = new Map<string, string>();
  for (const field of error.fieldErrors) {
    if (!field.path) continue;
    const fieldCode = field.code ?? 'usecase.input_invalid';
    fields.set(field.path, messages.t(fieldCode, field.params ?? {}));
  }

  const isServerFault = error.kind !== 'problem' || (error.status ?? 0) >= 500;
  const reference = error.requestId && !message.includes(error.requestId) ? error.requestId : undefined;

  return { message, fields, reference, isServerFault };
}
